lib/vtls/schannel.c

```diff
@@ -405,6 +405,362 @@ get_cert_location(TCHAR *path, DWORD *store_name, TCHAR **store_path,
+        SCH_CRED_IGNORE_REVOCATION_OFFLINE;
+
+      DEBUGF(infof(data, "schannel: disabled server certificate revocation "
+                   "checks\n"));
+    }
+    else if(data->set.ssl.revoke_best_effort) {
+      schannel_cred.dwFlags |= SCH_CRED_IGNORE_NO_REVOCATION_CHECK |
+        SCH_CRED_IGNORE_REVOCATION_OFFLINE | SCH_CRED_REVOCATION_CHECK_CHAIN;
+
+      DEBUGF(infof(data, "schannel: ignore revocation offline errors"));
+    }
+    else {
+      schannel_cred.dwFlags |= SCH_CRED_REVOCATION_CHECK_CHAIN;
+
+      DEBUGF(infof(data,
+                   "schannel: checking server certificate revocation\n"));
+    }
+  }
+  else {
+    schannel_cred.dwFlags = SCH_CRED_MANUAL_CRED_VALIDATION |
+      SCH_CRED_IGNORE_NO_REVOCATION_CHECK |
+      SCH_CRED_IGNORE_REVOCATION_OFFLINE;
+    DEBUGF(infof(data,
+                 "schannel: disabled server cert revocation checks\n"));
+  }
+
+  if(!conn->ssl_config.verifyhost) {
+    schannel_cred.dwFlags |= SCH_CRED_NO_SERVERNAME_CHECK;
+    DEBUGF(infof(data, "schannel: verifyhost setting prevents Schannel from "
+                 "comparing the supplied target name with the subject "
+                 "names in server certificates.\n"));
+  }
+
+  switch(conn->ssl_config.version) {
+  case CURL_SSLVERSION_DEFAULT:
+  case CURL_SSLVERSION_TLSv1:
+  case CURL_SSLVERSION_TLSv1_0:
+  case CURL_SSLVERSION_TLSv1_1:
+  case CURL_SSLVERSION_TLSv1_2:
+  case CURL_SSLVERSION_TLSv1_3:
+    {
+      result = set_ssl_version_min_max(&schannel_cred, data, conn);
+      if(result != CURLE_OK)
+        return result;
+      break;
+    }
+  case CURL_SSLVERSION_SSLv3:
+    schannel_cred.grbitEnabledProtocols = SP_PROT_SSL3_CLIENT;
+    break;
+  case CURL_SSLVERSION_SSLv2:
+    schannel_cred.grbitEnabledProtocols = SP_PROT_SSL2_CLIENT;
+    break;
+  default:
+    failf(data, "Unrecognized parameter passed via CURLOPT_SSLVERSION");
+    return CURLE_SSL_CONNECT_ERROR;
+  }
+
+  if(SSL_CONN_CONFIG(cipher_list)) {
+    result = set_ssl_ciphers(&schannel_cred, SSL_CONN_CONFIG(cipher_list));
+    if(CURLE_OK != result) {
+      failf(data, "Unable to set ciphers to passed via SSL_CONN_CONFIG");
+      return result;
+    }
+  }
+
+#ifdef HAS_CLIENT_CERT_PATH
+  /* client certificate */
+  if(data->set.ssl.primary.clientcert) {
+    DWORD cert_store_name;
+    TCHAR *cert_store_path = NULL;
+    TCHAR *cert_thumbprint_str;
+    CRYPT_HASH_BLOB cert_thumbprint;
+    BYTE cert_thumbprint_data[CERT_THUMBPRINT_DATA_LEN];
+    HCERTSTORE cert_store;
+    FILE *fInCert = NULL;
+
+    TCHAR *cert_path = curlx_convert_UTF8_to_tchar(
+                           data->set.ssl.primary.clientcert);
+    if(!cert_path)
+      return CURLE_OUT_OF_MEMORY;
+
+    result = get_cert_location(cert_path, &cert_store_name,
+                               &cert_store_path, &cert_thumbprint_str);
+    if((result != CURLE_OK) && (data->set.ssl.primary.clientcert[0]!='\0'))
```

Comment:
Is it guaranteed that clientcert is initialized?

Reply:
Existing code, there could be a bug, there.